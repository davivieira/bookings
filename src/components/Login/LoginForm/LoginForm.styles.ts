import { Box, Button, styled } from "@mui/material";

export const SCContainer = styled(Box)(({ theme: { palette } }) => ({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  background: palette.background.default,
  padding: 20,
  gap: 20,
  borderRadius: 10
}))

export const SCLoginButton = styled(Button)(({ theme: { palette } }) => ({
  width: '100%',
  height: 50,
  color: palette.background.default,
  backgroundColor: palette.secondary.main,
  '&:hover': {
    backgroundColor: palette.background.default,
    color: palette.secondary.main
  }
}))