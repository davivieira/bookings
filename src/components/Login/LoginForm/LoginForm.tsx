import { useFormik } from "formik";
import * as Yup from "yup";
import { SCForm } from "@/styles";
import { SCContainer, SCLoginButton } from "./LoginForm.styles";
import TextField from "@mui/material/TextField";
import CircularProgress from "@mui/material/CircularProgress";
import usePostUserData from "../../../api/users/users";

const validationSchema = Yup.object({
  email: Yup.string().email("Invalid email"),
});

type LoginFormProps = {
  handleClose: () => void;
};

function LoginForm({ handleClose }: LoginFormProps) {
  const { mutate, isLoading } = usePostUserData();

  const {
    handleSubmit,
    handleChange,
    values,
    errors,
    touched,
    isSubmitting,
  } = useFormik({
    initialValues: {
      email: "",
    },
    validationSchema,
    onSubmit: async (values) => {
      mutate(values.email);
      handleClose();
    },
  });

  return (
    <SCContainer>
      <SCForm onSubmit={handleSubmit}>
        <TextField
          value={values.email}
          onChange={handleChange}
          placeholder="Use your email to login"
          error={touched.email && !!errors.email}
          name="email"
          type="email"
        />
        <SCLoginButton data-testid="submit-login-button" type="submit">
          {isSubmitting || isLoading ? <CircularProgress /> : "Login"}
        </SCLoginButton>
      </SCForm>
    </SCContainer>
  );
}

export default LoginForm;
