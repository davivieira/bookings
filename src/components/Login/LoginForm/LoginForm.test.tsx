import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import LoginForm from './LoginForm';
import usePostUserData from '../../../api/users/users';
import { QueryClient, QueryClientProvider } from 'react-query';

jest.mock('../../../api/users/users');

const mockUsePostUserData = usePostUserData as jest.Mock;

describe('LoginForm', () => {
  const queryClient = new QueryClient();

  const renderWithClient = (ui: React.ReactElement) => {
    return render(
      <QueryClientProvider client={queryClient}>
        {ui}
      </QueryClientProvider>
    );
  };

  it('renders the login form and submits correctly', async () => {
    const handleClose = jest.fn();
    const mockMutate = jest.fn();
    mockUsePostUserData.mockReturnValue({ mutate: mockMutate, isLoading: false });

    renderWithClient(<LoginForm handleClose={handleClose} />);

    const emailInput = screen.getByPlaceholderText('Use your email to login');
    const submitButton = screen.getByRole('button', { name: /login/i });

    await userEvent.type(emailInput, 'test@example.com');
    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(mockMutate).toHaveBeenCalledWith('test@example.com');
      expect(handleClose).toHaveBeenCalled();
    });
  });

  it('shows loading indicator when submitting', async () => {
    const handleClose = jest.fn();
    const mockMutate = jest.fn();
    mockUsePostUserData.mockReturnValue({ mutate: mockMutate, isLoading: true });

    renderWithClient(<LoginForm handleClose={handleClose} />);

    const emailInput = screen.getByPlaceholderText('Use your email to login');
    const submitButton = screen.getByRole('button');

    await userEvent.type(emailInput, 'test@example.com');
    await userEvent.click(submitButton);

    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const handleClose = jest.fn();
    mockUsePostUserData.mockReturnValue({ mutate: jest.fn(), isLoading: false });

    const { asFragment } = renderWithClient(<LoginForm handleClose={handleClose} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
