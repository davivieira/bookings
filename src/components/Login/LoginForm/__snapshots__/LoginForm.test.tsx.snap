// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-p9ca7l"
  >
    <form
      class="css-ybf384"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            name="email"
            placeholder="Use your email to login"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-12690t6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Login
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</DocumentFragment>
`;
